kong/dns/client.lua

```diff
@@ -0,0 +1,691 @@
+      return answers, nil, answers.ttl
+    end
+  end
+
+  -- not found in the search iteration
+  return answers, nil, answers.ttl
+end
+
+
+local function resolve_query_types(self, name, qtype, tries)
+  local names = search_names(name, self.resolv, self.hosts)
+  local answers, err, ttl
+
+  -- the specific type
+  if qtype ~= TYPE_A_AAAA then
+    return resolve_query_names(self, names, qtype, tries)
+  end
+
+  -- query A or AAAA
+  if self.enable_ipv4 then
+    answers, err, ttl = resolve_query_names(self, names, TYPE_A, tries)
+    if not answers or not answers.errcode then
+      return answers, err, ttl
+    end
+  end
+
+  if self.enable_ipv6 then
+    answers, err, ttl = resolve_query_names(self, names, TYPE_AAAA, tries)
+  end
+
+  return answers, err, ttl
+end
+
+
+local function stale_update_task(premature, self, key, name, qtype)
+  if premature then
+    return
+  end
+
+  local tries = setmetatable({}, _TRIES_MT)
+  local answers = resolve_query_types(self, name, qtype, tries)
+  if not answers or answers.errcode then
+    log(DEBUG, PREFIX, "failed to update stale DNS records: ", tostring(tries))
+    return
+  end
+
+  log(DEBUG, PREFIX, "update stale DNS records: ", #answers)
+  self.cache:set(key, { ttl = answers.ttl }, answers)
+end
+
+
+local function start_stale_update_task(self, key, name, qtype)
+  stats_increment(self.stats, key, "stale")
+
+  local ok, err = timer_at(0, stale_update_task, self, key, name, qtype)
+  if not ok then
+    log(ALERT, PREFIX, "failed to start a timer to update stale DNS records: ", err)
+  end
+end
+
+
+local function check_and_get_ip_answers(name)
+  -- TODO: use is_valid_ipv4 from kong/tools/ip.lua instead
+  if name:match("^%d+%.%d+%.%d+%.%d+$") then  -- IPv4
+    return {
+      { name = name, class = 1, type = TYPE_A, address = name },
+    }
+  end
+
+  if name:find(":", 1, true) then             -- IPv6
+    return {
+      { name = name, class = 1, type = TYPE_AAAA, address = ipv6_bracket(name) },
+    }
+  end
+
+  return nil
+end
+
+
+local function resolve_callback(self, name, qtype, cache_only, tries)
+  -- check if name is ip address
+  local answers = check_and_get_ip_answers(name)
+  if answers then -- domain name is IP literal
+    answers.ttl = LONG_LASTING_TTL
+    answers.expire = now() + answers.ttl
+    return answers, nil, answers.ttl
+  end
+
+  -- check if this key exists in the hosts file (it maybe evicted from cache)
+  local key = name .. ":" .. qtype
+  local answers = self.hosts_cache[key]
+  if answers then
+    return answers, nil, answers.ttl
+  end
+
+  -- `:peek(stale=true)` verifies if the expired key remains in L2 shm, then
+  -- initiates an asynchronous background updating task to refresh it.
+  local ttl, _, answers = self.cache:peek(key, true)
+  if answers and not answers.errcode and ttl then
+    if not answers.expired then
+      answers.expire = now() + ttl
+      answers.expired = true
+      ttl = ttl + self.stale_ttl
```

Comment:
```suggestion
      remaining_stale_ttl = ttl + self.stale_ttl
```

Also, the word `ttl` is a bit abused here and this logic is hard to follow (took me a while to wrap my head around). Can we add some comment explaining what the calculations is doing here?